runtime/jvmti/jvmtiThread.c

```diff
@@ -391,21 +391,39 @@ jvmtiStopThread(jvmtiEnv *env,
 
 		rc = getVMThread(
 				currentThread, thread, &targetThread,
-#if JAVA_SPEC_VERSION >= 19
-				JVMTI_ERROR_UNSUPPORTED_OPERATION,
-#else /* JAVA_SPEC_VERSION >= 19 */
 				JVMTI_ERROR_NONE,
-#endif /* JAVA_SPEC_VERSION >= 19 */
-				J9JVMTI_GETVMTHREAD_ERROR_ON_NULL_JTHREAD | J9JVMTI_GETVMTHREAD_ERROR_ON_DEAD_THREAD | J9JVMTI_GETVMTHREAD_ERROR_ON_VIRTUALTHREAD);
+				J9JVMTI_GETVMTHREAD_ERROR_ON_NULL_JTHREAD | J9JVMTI_GETVMTHREAD_ERROR_ON_DEAD_THREAD);
 		if (JVMTI_ERROR_NONE == rc) {
+#if JAVA_SPEC_VERSION >= 21
+			j9object_t threadObject = (NULL == thread) ? currentThread->threadObject : J9_JNI_UNWRAP_REFERENCE(thread);
+
+			/* Error if a virtual thread is not suspended and not the current thread. */
+			if ((currentThread != targetThread)
+			&& (0 == J9OBJECT_U32_LOAD(currentThread, threadObject, vm->isSuspendedInternalOffset))
+			&& IS_JAVA_LANG_VIRTUALTHREAD(currentThread, threadObject)
+			) {
+				rc = JVMTI_ERROR_THREAD_NOT_SUSPENDED;
+				goto release;
+			}
+
+			/* Error if a virtual thread is unmounted since it won't be able to throw an
+			 * asynchronous exception from the current frame.
+			 */
+			if (NULL == targetThread) {
+				rc = JVMTI_ERROR_OPAQUE_FRAME;
+				goto release;
+			}
+#endif /* JAVA_SPEC_VERSION >= 21 */
 			omrthread_monitor_enter(targetThread->publicFlagsMutex);
 			if (OMR_ARE_NO_BITS_SET(targetThread->publicFlags, J9_PUBLIC_FLAGS_STOPPED)) {
 				omrthread_priority_interrupt(targetThread->osThread);
 				targetThread->stopThrowable = J9_JNI_UNWRAP_REFERENCE(exception);
-				clearHaltFlag(targetThread, J9_PUBLIC_FLAGS_HALT_THREAD_JAVA_SUSPEND);
```

Comment:
[StopThreadTest](https://github.com/ibmruntimes/openj9-openjdk-jdk21/blob/737f940cd67989a5eeb8dcb79c94a1d2d7266b27/test/hotspot/jtreg/serviceability/jvmti/vthread/StopThreadTest/StopThreadTest.java) has two modes. One where it runs with platform/normal threads and the other where it runs with virtual threads. The same error is seen in both modes:

> `JVMTI_ERROR_THREAD_NOT_SUSPENDED` is returned from `ResumeThread` because `StopThread` has cleared the `J9_PUBLIC_FLAGS_HALT_THREAD_JAVA_SUSPEND` flag when the following sequence of JVMTI functions are
invoked: `SuspendThread -> StopThread -> ResumeThread`.

So, I believe that the above change is needed for all cases.

Reply:
The "spec" for StopThread is a single sentence, so it's difficult to base anything on that. Need to verify the behaviour on a suspended normal thread (and do the same on Java 8 to see if the behaviour varies by java level).

Reply:
Confirming that the same error is also seen with OpenJ9 Java 8; removing `clearHaltFlag(...)` fixes the error and matches the RI behaviour for both Java 8 and Java 21.

Rewrote [StopThreadTest](https://github.com/ibmruntimes/openj9-openjdk-jdk21/blob/737f940cd67989a5eeb8dcb79c94a1d2d7266b27/test/hotspot/jtreg/serviceability/jvmti/vthread/StopThreadTest/StopThreadTest.java) so that it can be run with Java 8 and Java 21: [jvmti_stoptest.zip](https://github.com/eclipse-openj9/openj9/files/12088487/jvmti_stoptest.zip). Now, it only uses normal/platform threads. Instructions to run the test:
```
wget https://github.com/eclipse-openj9/openj9/files/12088487/jvmti_stoptest.zip

unzip -q jvmti_stoptest.zip

cd jvmti_stoptest

source run.sh <PATH_TO_JAVA_HOME>
```